src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
+
+        SuccessOrExit(err = SendSigma2());
+    }
 
 exit:
 
-    if (err != CHIP_NO_ERROR)
+    if (err == CHIP_ERROR_CERT_NOT_TRUSTED)
+    {
+        SendStatusReport(mExchangeCtxt, kProtocolCodeNoSharedRoot);
+        mState = kInitialized;
+    }
+    else if (err != CHIP_NO_ERROR)
     {
-        SendErrorMsg(SigmaErrorType::kUnexpected);
+        SendStatusReport(mExchangeCtxt, kProtocolCodeInvalidParam);
+        mState = kInitialized;
     }
     return err;
 }
 
-CHIP_ERROR CASESession::SendSigmaR2()
+CHIP_ERROR CASESession::SendSigma2Resume(const ByteSpan & initiatorRandom)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    size_t data_len = EstimateTLVStructOverhead(kCASEResumptionIDSize + kTAGSize + sizeof(uint16_t), 7);
 
-    System::PacketBufferHandle msg_R2;
-    size_t data_len;
+    System::PacketBufferTLVWriter tlvWriter;
+    System::PacketBufferHandle msg_R2_resume;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
 
-    uint8_t msg_rand[kSigmaParamRandomNumberSize];
+    msg_R2_resume = System::PacketBufferHandle::New(data_len);
+    VerifyOrReturnError(!msg_R2_resume.IsNull(), CHIP_ERROR_NO_MEMORY);
 
-    chip::Platform::ScopedMemoryBuffer<uint8_t> msg_R2_Signed;
-    size_t msg_r2_signed_len;
+    tlvWriter.Init(std::move(msg_R2_resume));
 
-    chip::Platform::ScopedMemoryBuffer<uint8_t> msg_R2_Encrypted;
-    size_t msg_r2_signed_enc_len;
+    // Generate a new resumption ID
+    ReturnErrorOnFailure(DRBG_get_bytes(mResumptionId, sizeof(mResumptionId)));
 
-    uint8_t msg_salt[kIPKSize + kSigmaParamRandomNumberSize + kP256_PublicKey_Length + kSHA256_Hash_Length];
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(1), mResumptionId, kCASEResumptionIDSize));
 
-    uint8_t sr2k[kAEADKeySize];
-    P256ECDSASignature tbsData2Signature;
+    chip::Platform::ScopedMemoryBuffer<uint8_t> sigma2ResumeMIC;
+    ReturnErrorCodeIf(!sigma2ResumeMIC.Alloc(kTAGSize), CHIP_ERROR_NO_MEMORY);
+    MutableByteSpan resumeMICSpan(sigma2ResumeMIC.Get(), kTAGSize);
+    ReturnErrorOnFailure(GenerateSigmaResumeMIC(initiatorRandom, ByteSpan(mResumptionId), ByteSpan(kKDFS2RKeyInfo),
+                                                ByteSpan(kResume2MIC_Nonce), resumeMICSpan));
 
-    ByteSpan icaCert;
-    ByteSpan nocCert;
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(2), resumeMICSpan));
 
-    VerifyOrExit(mFabricInfo != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(3), GetLocalSessionId(), true));
 
-    SuccessOrExit(err = mFabricInfo->GetICACert(icaCert));
-    SuccessOrExit(err = mFabricInfo->GetNOCCert(nocCert));
+    // TODO: Add support for optional MRP parameters
+
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&msg_R2_resume));
+
+    mState = kSentSigma2Resume;
```

Comment:
This should be set after succesful send

Reply:
This had to be done here because in unit-test harness, `SendMessage()` is directly calling `OnMessageReceived()`. So the message handler for the received message gets looked up before the SendMessage() would return. If the state is not already advanced to the next state, the handler lookup will refer to old state, and will fail.

To account for SendMessage() errors, I am resetting the state to kInitialized in all such scenarios.

Reply:
This couples the code to the test in an open-box fashion that is not scalable. The processing of messages should be done such that OnMessageReceived is never called prior to finishing the handling of a send, since that would never happen on a real product and could lead to unrealistic timing scenarios. If this is truly required for now, there needs to be a clear comment of the fact that this is done before the SendMessage to support current testing code.

Reply:
Added TODO to update unit test harness.